src/librustc/traits/project.rs

```diff
@@ -818,55 +819,57 @@ fn project_type<'cx, 'gcx, 'tcx>(
 
+    let decide_commit = |candidates: &mut ProjectionTyCandidateSet<'tcx>| {
+        debug!("{} candidates, ambiguous={}",
+               candidates.vec.len(),
+               candidates.ambiguous);
+
+        // Prefer where-clauses. As in select, if there are multiple
+        // candidates, we prefer where-clause candidates over impls.  This
+        // may seem a bit surprising, since impls are the source of
+        // "truth" in some sense, but in fact some of the impls that SEEM
+        // applicable are not, because of nested obligations. Where
+        // clauses are the safer choice. See the comment on
+        // `select::SelectionCandidate` and #21974 for more details.
+        if candidates.vec.len() > 1 {
+            debug!("retaining param-env candidates only from {:?}", candidates.vec);
+            candidates.vec.retain(|c| match *c {
+                ProjectionTyCandidate::ParamEnv(..) => true,
+                ProjectionTyCandidate::TraitDef(..) |
+                ProjectionTyCandidate::Select(..) => false,
+            });
+            debug!("resulting candidate set: {:?}", candidates.vec);
+            if candidates.vec.len() != 1 {
+                candidates.ambiguous = true;
+                return false;
+            }
+        }
+
+        assert!(candidates.vec.len() <= 1);
+        match candidates.vec.get(0) {
+            Some(&ProjectionTyCandidate::Select(..)) => true,
+            _ => false,
+        }
+    };
+
+    // Note that at here, if `ProjectionTyCandidate::Select` is not going to be
+    // a valid candidate, the closure is not executed at all. There are two cases,
+    // one being trait selection error, and another being ambiguous candidates.
+    // We handle both by the early return below.
     if let Err(e) = assemble_candidates_from_impls(selcx,
                                                    obligation,
                                                    &obligation_trait_ref,
-                                                   &mut candidates) {
+                                                   &mut candidates,
+                                                   decide_commit) {
         return Err(ProjectionTyError::TraitSelectionError(e));
     }
 
-    debug!("{} candidates, ambiguous={}",
-           candidates.vec.len(),
-           candidates.ambiguous);
-
     // Inherent ambiguity that prevents us from even enumerating the
     // candidates.
     if candidates.ambiguous {
         return Err(ProjectionTyError::TooManyCandidates);
     }
 
-    // Drop duplicates.
-    //
-    // Note: `candidates.vec` seems to be on the critical path of the
-    // compiler. Replacing it with an HashSet was also tried, which would
-    // render the following dedup unnecessary. The original comment indicated
-    // that it was 9% slower, but that data is now obsolete and a new
-    // benchmark should be performed.
-    candidates.vec.sort_unstable();
-    candidates.vec.dedup();
```

Comment:
Indeed I can't really see how duplicates would appear in here right now -- unless there were duplicates in the param-env etc, but the elaboration step removes those right now, so I think that shouldn't happen. Maybe add an assertion in the `push_candidate` step above -- when we push a candidate but still have an old one, they should never be the same.